src/core/numerics/pusher/pusher.h

```diff
@@ -3,19 +3,23 @@
 
 #include <cstddef>
 #include <utility>
+#include <functional>
+
 #include "core/utilities/range/range.h"
+#include "core/data/particles/particle.h"
 
 namespace PHARE
 {
 namespace core
 {
     template<std::size_t dim, typename ParticleIterator, typename Electromag, typename Interpolator,
-             typename ParticleSelector, typename BoundaryCondition, typename GridLayout>
+             typename BoundaryCondition, typename GridLayout>
     class Pusher
     {
     protected:
-        using ParticleRange = Range<ParticleIterator>;
-
+        using ParticleRange             = Range<ParticleIterator>;
+        static auto constexpr dimension = GridLayout::dimension;
+        using ParticleSelector          = std::function<bool(Particle<dimension> const&)>;
```

Comment:
is the class redundant now?

Reply:
ah yeah, I wanted to remove it and forgot.
Yeah, so the reason behind the lambdas is that:

- first, it's simpler to define a lambda than a functor
- second it's also more adequate because you'll notice that the class ParticleSelector basically assume you're going to select from a box (or a collection of, as long as the criteria is "isIn"). But now in the IonUpdater, when we push patchGhost particles, we want a selector that checks whether they leave the patch ghost area, which is basically "in ghost box but not in domain box".... ParticleSelector didn't work for that, but lambdas can do whatever you want.... so simpler.

I'll remove the ParticleSelector class